api/core/prompt/agent_history_prompt_transform.py

```diff
@@ -0,0 +1,165 @@
+                    tools = agent_thought.tool
+                    if tools:
+                        tools = tools.split(';')
+                        tool_calls: list[AssistantPromptMessage.ToolCall] = []
+                        tool_call_response: list[ToolPromptMessage] = []
+                        try:
+                            tool_inputs = json.loads(agent_thought.tool_input)
+                        except Exception as e:
+                            tool_inputs = {tool: {} for tool in tools}
+                        try:
+                            tool_responses = json.loads(agent_thought.observation)
+                        except Exception as e:
+                            tool_responses = {tool: agent_thought.observation for tool in tools}
+
+                        for tool in tools:
+                            # generate a uuid for tool call
+                            tool_call_id = str(uuid.uuid4())
+                            tool_calls.append(AssistantPromptMessage.ToolCall(
+                                id=tool_call_id,
+                                type='function',
+                                function=AssistantPromptMessage.ToolCall.ToolCallFunction(
+                                    name=tool,
+                                    arguments=json.dumps(tool_inputs.get(tool, {})),
+                                )
+                            ))
+                            tool_call_response.append(ToolPromptMessage(
+                                content=tool_responses.get(tool, agent_thought.observation),
+                                name=tool,
+                                tool_call_id=tool_call_id,
+                            ))
+
+                        prompt_messages.extend([
+                            AssistantPromptMessage(
+                                content=agent_thought.thought,
+                                tool_calls=tool_calls,
+                            ),
+                            *tool_call_response
+                        ])
+                        num_prompt += 1 + len(tool_call_response)
+                    if not tools:
+                        prompt_messages.append(AssistantPromptMessage(content=agent_thought.thought))
+                        num_prompt += 1
+            else:
+                if message.answer:
+                    prompt_messages.append(AssistantPromptMessage(content=message.answer))
+                    num_prompt += 1
+
+            curr_message_tokens = model_type_instance.get_num_tokens(
+                self.memory.model_instance.model,
+                self.memory.model_instance.credentials,
+                prompt_messages
+            )
+            # If tokens is overflow, drop all appended prompts in current message and break
+            if curr_message_tokens > max_token_limit:
+                prompt_messages = prompt_messages[:-num_prompt]
+                break
```

Comment:
hi, looks like this logic will break out the loop and only remove messages of current agent thought? why not remove first N message instead of last N, BTW, a single iteration could produces several messages, and a single agent process may have multiple iterations, we need to solve this inside agent runner before every iteration I think.

Reply:
Of course, you can keep your idea and I will fix them later, I will approve it if you don't mind : )

Reply:
The order of messages is desc, it will remove the first N messages. 
It does has problem in multiple iterations. I think organize the prompt messages before every iteration may solve the problem. I'm working on it.

Reply:
Oh, I see that you're using descending order, great observation! But this will reverse the entire message history. I haven't noticed any logic to revert it back. Is this intentional or a mistake?

Reply:
Yes, it's a mistake, history messages should be reorganized in ascending order with recalculating tokens. I will fix it.